### PR TITLE
idprime: Add known ATRs and card owners

### PR DESCRIPTION
**Pull request by Jakuje**
I would like to keep a list of know ATRs for different IDPrime cards that are known to work to avoid issues such as #3106 in the future.

I added mine cards + cards from #3106 and what I was able to find in the old pull requests. If anyone else has another IDPrime, please comment here so I can add it (or create a PR with changes).

##### Checklist

- [X] Documentation is added or updated
- [X] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

**Comment by Jakuje** [comment]
Already asked @Nihlus in previous PR, would like to hear from @jurajsarinay @kirichkov or anyone who has this type of card.

**Comment by Nihlus** [comment]
Continuing on; I've reached out to see about getting the ATRs. Hopefully we can get a sample set!

**Comment by metsma** [comment]
I have also available
Product name: SafeNet eToken 5110 CC
Card type: IDPrime
Applet Version: IDPrime Java Applet 4.0.2.K
Mask version: G260/Filt17
FIPS Profile: N/A
Common Criteria (CC) Profile: CC EAL5+ / PP QSCD 
Sign padding on-board: Yes
Supported key size: 4096 bits
ECC: Supported
3b:ff:96:00:00:81:31:80:43:80:31:80:65:b0:85:03:00:ef:12:0f:fe:82:90:00:66

and
Product name: eToken 5110 CC (940)
Card type: IDPrime
Applet Version: IDPrime Java Applet 4.4.2.A
FIPS Profile: N/A
Common Criteria (CC) Profile: CC EAL5+ / PP QSCD 
Sign padding on-board: Yes
Supported key size: 4096 bits
ECC: Supported
3b:ff:96:00:00:81:31:fe:43:80:31:80:65:b0:85:59:56:fb:12:0f:fe:82:90:00:00

**Comment by Jakuje** [comment]
Thank you! Added them to the list.

The information you provided suggests that these cards should support both RSA 4k and ECC, but the driver matches them as SC_CARD_TYPE_IDPRIME_GENERIC (correct me if I did the matching wrongly by hand), which does not turn these features on. But the ATRs are very similar to the one `3b:ff:96:00:00:81:31:fe:43:80:31:80:65:b0:84:65:66:fb:12:01:78:82:90:00:85` which is matched as `SC_CARD_TYPE_IDPRIME_930`. Do all of them work for you now with current master with all features?

This would be probably a question for Raul to check. The ` Product name: eToken 5110 CC (940)` sounds like a reference to IDPrime 940, but I would need some testing if the card behaves similarly like our idprimes in regards to the indexes, key references and so on.

**Comment by Jakuje** [comment]
Thank you! So my prerefence would be to move your ATR to the more specific class `SC_CARD_TYPE_IDPRIME_940`. But it completely overlaps with the ATR `3b:ff:96:00:00:81:31:80:43:80:31:80:65:b0:85:03:00:ef:12:0f:fe:82:90:00:66` from @metsma but the names look different in the output. Can you check if your card with this ATR works when matched as either of the non-generic ones?

**Comment by kirichkov** [comment]
The one I have left has atr `3b:7f:96:00:00:80:31:80:65:b0:85:59:56:fb:12:0f:fe:82:90:00`

File: src/libopensc/card-idprime.c [hunk 1]
```diff
@@ -45,41 +45,72 @@ static struct sc_card_driver idprime_drv = {
...
 	  "Gemalto IDPrime 940C",
-	  SC_CARD_TYPE_IDPRIME_940, 0, NULL },	
+	  SC_CARD_TYPE_IDPRIME_940, 0, NULL },
+	/* Known ATRs for IDPrime 940 (eToken 5110)
+	 * 3b:ff:96:00:00:81:31:fe:43:80:31:80:65:b0:85:59:56:fb:12:0f:fe:82:90:00:00    metsma, jurajsarinay
+	 */
+	{ "3b:ff:96:00:00:81:31:fe:43:80:31:80:65:b0:85:59:56:fb:12:0f:fe:82:90:00:00",
+	  "ff:ff:00:ff:ff:ff:ff:00:ff:ff:ff:ff:ff:ff:00:00:00:00:ff:ff:ff:ff:ff:ff:00",
+	  "Gemalto IDPrime MD 940 (eToken 5110)",
+	  SC_CARD_TYPE_IDPRIME_GENERIC, 0, NULL },
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/libopensc/card-idprime.c]: Indeed! Thanks for double-checking. Fixed

**Review by frankmorgner — COMMENTED** [review]
I would rather avoid creating an inventory inside the source code. Maybe the wiki is a better place to put people with their known cards. Maybe a dedicated test file in the repository would work as well.

**Comment by Jakuje** [comment]
> I would rather avoid creating an inventory inside the source code. Maybe the wiki is a better place to put people with their known cards. Maybe a dedicated test file in the repository would work as well.

My reasoning was to have all the existing ATRs aligned next time when we will get a new ATR, request to change the masks or anything to see the results of such changes on existing users and to be able to catch mismatches earlier.

If it will be in Wiki, we would need to compare the ATRs from wiki with the masks and matches in code or keeping them in sync, which is hard. But keeping them in unit test might work. Let me see if I can put something together

**Comment by Nihlus** [comment]
@Jakuje My colleague has pulled the ATR now :)
https://github.com/OpenSC/OpenSC/pull/2958#issuecomment-2188897747

**Comment by dengert** [comment]
> If it will be in Wiki, we would need to compare the ATRs from wiki with the masks and matches in code or keeping them in sync, which is hard. But keeping them in unit test might work. Let me see if I can put something together

That could also be used to check if multiple drivers (with masks) might test or claim the card as supported by the driver. Last thing we want to do is add an ATR and mask that causes some other driver to fail.

**Comment by ebababi** [comment]
Hi @jakuje and team, thank you for your great work. Please accept my 2 cents.

> So my prerefence would be to move your ATR to the more specific class SC_CARD_TYPE_IDPRIME_940.

I have SafeNet eToken 5110 CC (MD-940) with ATR:
```sh
$ opensc-tool -a
Using reader with a card: SafeNet eToken 5100 [eToken 5110 SC] 00 00
3b:ff:96:00:00:81:31:fe:43:80:31:80:65:b0:85:59:56:fb:12:0f:fe:82:90:00:00
```

When identified as `SC_CARD_TYPE_IDPRIME_GENERIC` I wasn't able to sign anything (no 2nd PIN), but when I applied your 
b853dfe patch identifying it as `SC_CARD_TYPE_IDPRIME_940` it worked by selecting the second slot. I think the reason that some people report success with the former configuration is that it supports what they call "LinkMode" - [the manual](https://aur.archlinux.org/packages/sac-core) writes "Use the same token and administrator passwords for digital signature operations.", however, this is not the default setting.

In any case, I think the proper identification is `SC_CARD_TYPE_IDPRIME_940`, as you suspected it.

**Comment by msetina** [comment]
I just got my hands on this:
Using reader with a card: Gemalto USB Shell Token V2 (67689F81) 00 00
3b:7f:96:00:00:80:31:80:65:b0:85:59:56:fb:12:02:c1:82:90:00
They say it contains IDPrime MD 940, it is not recognized.

**Comment by Jakuje** [comment]
@msetina did you try with this PR or with master? This PR also adds some reordering of ATRs to fix some of the bugs of ATR matching, but I did not manage to follow-up on this yet to get this merged.

I would certainly like to merge this in some form for 0.26.0. I agree with @frankmorgner concerns about having user names in source code, but I did not find any better way to be able to simply manually comparing the ATRs against the maps we have in the source code without the need to maintain the maps in two places (wiki + source code).

For all, there is one fix for IDPrime 940 in #3220, but I do not think it will resolve the issue you are seeing, but double-checking the changes do not break either of the cards anyone has would be great.

**Comment by msetina** [comment]
@Jakuje I have been working with 0.25.1 release + my changes. Right now I am not as far as the #3220 as I understand. As stated in previous comment the problem is with the iso file (idprime_select_file_by_path) request that get reply file not found. It looks like IDPrime MD 940 is not the same as IDPrime 940 as this code is expecting something that is not there. 
I can try with this PR, but I can see that there are duplicate ATRs in the PR and from my runs of similar code on a list like that and algo like that I got 940C match to 940 that is a row before 940C with more restrictive map. (The ATR for 940C matches the row of 940 (ATR+map))

**Comment by xhanulik** [comment]
This card is not known. Even adding proper entry to table, I get generic driver. Trying to debug I found out that answer to APDU:
> 
> ```
> 00 A4 04 0C 0B A0 00 00 00 18 80 00 00 00 06 62 ...............b
> ```
> 
> is:
> 
> ```
> 6A 82 j.
> ```


@msetina Could you please provide OpenSC debug log where the driver matching part is visible? I think we haven't encountered IDPrime with different AID so far.

**Comment by msetina** [comment]
@xhanulik Here it is:
[MD_940_test_940.txt](https://github.com/user-attachments/files/16712216/MD_940_test_940.txt)

I would also comment that the procedure moved to a new driver even tho the match was done, but did not get proper FS.

**Comment by xhanulik** [comment]
@msetina Thank you, it seems that regardless of the ATR order, it won't match just because the selecting AID by idprime driver returns _file is not found_. Do you have some proprietary IDPrime PKCS#11 library, so you would be able to create some log from it? Some steps how it can be done are described in https://github.com/OpenSC/OpenSC/issues/3201#issuecomment-2239012515, it can be whatever command, just to see what AID is actually used and working.

**Comment by msetina** [comment]
@xhanulik sent to email in your profile. First APDU is longer, but there is some more chater after compared to using normal 940 for the same command.

**Comment by xhanulik** [comment]
@msetina I went through the log.

First thing, at the beggining it selects different AID than the AID used for IDPrime in OpenSC driver. There is also reading of files, which are not used in IDPrime driver. I haven't seen IDPrime card with different ATR yet, but at least we can try to add your AID to the driver and check, whether rest of the IDPrime stuff would work then.
I have also found the ATR `3B 7F 96 00 00 80 31 80 65 B0 85 59 56 FB 12 02 C1 82 90 00` above listed in https://pcsc-tools.apdu.fr/smartcard_list.txt as `Gemalto USB (eID)` but nothing else regarding the applet.
For the log you sent me, what library did you use? The `libIDPrimePKCS11.so` or something different?

I would maybe suggest to move this into separate issue, because it might not have such a straightforward solution.

**Comment by msetina** [comment]
Yes in my case it is Gemalto USB, but that is a reader. The same provider also has cards which are personalized. The card in the reader is identified as IdPrime MD940 and they also say it is a cutout. Unfortunately the driver reports that it can only generate keys with 2048 bits. There are some operations that it reports that it can use 4096 bits. 
I can use the card with Nexus Personal Client I got from provider.

**Comment by xhanulik** [comment]
@metsma That is interesting. I will add the AID to try it with the IDPrime driver. When you have time, please create a follow-up issue.

Also I think, storing 2-line comments in the ATR table for the driver is still okay, as there are other drivers (such as PIV) that also have some additional info and, in my opinion, are more usable than storing them in a separate wiki page.

Otherwise, going through the conversation here, this PR should be in a state that works fine for all the mentioned cards.

**Review by xhanulik — APPROVED** [review]

**Comment by xhanulik** [comment]
Merging, thank you!

**Pull request merged**